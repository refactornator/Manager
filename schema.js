export default (schema = {
  schema: [
    {
      name: 'Note',
      primaryKey: 'key',
      properties: {
        key: 'string',
        text: 'string',
        createdAt: 'date',
        report: { type: 'Report', optional: true }
      }
    },
    {
      name: 'Report',
      primaryKey: 'key',
      properties: {
        key: 'string',
        name: 'string',
        notes: { type: 'list', objectType: 'Note' }
      }
    }
  ]
});
