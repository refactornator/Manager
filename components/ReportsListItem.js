import {
  AlertIOS,
  Dimensions,
  LayoutAnimation,
  TouchableOpacity
} from 'react-native';
import Icon from 'react-native-vector-icons/FontAwesome';
import React, { Component } from 'react';
import styled from 'styled-components/native';
import uuid from 'uuid';

const { width: windowWidth } = Dimensions.get('window');

export default class ReportsListItem extends Component {
  addReport = name => {
    const { realm } = this.props;

    try {
      realm.write(() => {
        const newReport = realm.create('Report', {
          key: uuid.v1(),
          name
        });
      });
      LayoutAnimation.configureNext(LayoutAnimation.Presets.spring);
    } catch (e) {
      console.log('Error on creation', e);
    }
  };

  render() {
    const { navigation, realm, data } = this.props;

    if (typeof data === 'object') {
      return (
        <TouchableOpacity onPress={() => navigation.navigate('Report', data)}>
          <Item>
            <Icon name="user" size={44} color="#366ddc" />
            <Name>{data.name}</Name>
          </Item>
        </TouchableOpacity>
      );
    } else if (data === 'add') {
      return (
        <TouchableOpacity
          onPress={() => {
            AlertIOS.prompt('Enter Name', null, this.addReport, 'plain-text');
          }}
        >
          <Item>
            <Icon name="plus-circle" size={44} color="#366ddc" />
            <Name>Add Report</Name>
          </Item>
        </TouchableOpacity>
      );
    }
  }
}

const Item = styled.View`
  flex: 1;
  width: ${windowWidth / 2}px;
  height: ${windowWidth / 2}px;
  align-items: center;
  justify-content: center;
`;

const Name = styled.Text`
  color: #366ddc;
  margin-top: 6px;
  font-size: 20px;
`;
