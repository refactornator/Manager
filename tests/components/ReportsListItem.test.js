import React from 'react';

import { shallow } from 'enzyme';
import sinon from 'sinon';

import Realm from '../mocks/realmMock';
import ReportsListItem from '../../components/ReportsListItem';
import schema from '../../schema';

const realm = new Realm(schema);
const data = [{ key: 1, name: 'William' }];
realm.prepareData('Report', data);

const navigation = { navigate: () => {} };

describe('report list item', () => {
  let render;
  beforeEach(() => {
    render = shallow(
      <ReportsListItem {...{ navigation, realm, data: data[0] }} />
    );
  });

  it('renders', () => {
    expect(render).toMatchSnapshot();
  });

  describe('when pressed', () => {
    it("navigates to the report's screen", () => {
      navigation.navigate = sinon.spy();

      const item = render.find('TouchableOpacity');
      expect(item.length).toBe(1);

      item.simulate('press');
      expect(navigation.navigate.calledWith('Report', data[0])).toBe(true);

      navigation.navigate = () => {};
    });
  });
});

describe('add report list item', () => {
  let render;
  beforeEach(() => {
    render = shallow(
      <ReportsListItem {...{ navigation, realm, data: 'add' }} />
    );
  });

  it('renders', () => {
    expect(render).toMatchSnapshot();
  });

  describe('when pressed', () => {
    it('opens the alert dialog', () => {
      expect(render.find('TouchableOpacity').length).toBe(1);
    });
  });
});
