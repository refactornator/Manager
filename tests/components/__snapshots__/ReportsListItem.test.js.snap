// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`add report list item renders 1`] = `
<TouchableOpacity
  activeOpacity={0.2}
  onPress={[Function]}
>
  <Styled(View)>
    <Icon
      allowFontScaling={false}
      color="#366ddc"
      name="plus-circle"
      size={44}
    />
    <Styled(Text)>
      Add Report
    </Styled(Text)>
  </Styled(View)>
</TouchableOpacity>
`;

exports[`report list item renders 1`] = `
<TouchableOpacity
  activeOpacity={0.2}
  onPress={[Function]}
>
  <Styled(View)>
    <Icon
      allowFontScaling={false}
      color="#366ddc"
      name="user"
      size={44}
    />
    <Styled(Text)>
      William
    </Styled(Text)>
  </Styled(View)>
</TouchableOpacity>
`;
